opentelemetry-exporter-geneva/geneva-uploader/src/config_service/client.rs

```diff
@@ -0,0 +1,536 @@
+        &self,
+    ) -> Result<(IngestionGatewayInfo, MonikerInfo, String)> {
+        // First, try to read from cache (shared read access)
+        {
+            let cached_data: std::sync::RwLockReadGuard<'_, CachedAuthData> =
+                self.cached_data.read().map_err(|_| {
+                    GenevaConfigClientError::InternalError("RwLock poisoned".to_string())
+                })?;
+
+            if let Some(expiry) = cached_data.token_expiry {
+                // Use 5-minute buffer for token expiry check
+                // Token lasts for 3 days (As observed), so 5 minutes is plenty of safety margin
+                // This ensures we have time to refresh before the token expires
+                if expiry > Utc::now() + chrono::Duration::minutes(5) {
+                    if let (Some((ingestion, moniker)), Some(endpoint)) =
+                        (&cached_data.auth_info, &cached_data.token_endpoint)
+                    {
+                        return Ok((ingestion.clone(), moniker.clone(), endpoint.clone()));
+                    }
+                }
+            }
+        }
+
+        // Cache miss or expired token, fetch fresh data
+        // Perform actual fetch before acquiring write lock to minimize lock contention
+        let (fresh_token_info, fresh_moniker_info) = self.fetch_ingestion_info().await?;
+        let token_expiry = self.parse_token_expiry(&fresh_token_info.auth_token_expiry_time);
+        let token_endpoint =
+            extract_endpoint_from_token(&fresh_token_info.auth_token).map_err(|e| {
+                GenevaConfigClientError::Certificate(format!(
+                    "Failed to extract endpoint from token: {}",
+                    e
+                ))
+            })?;
+
+        // Now update the cache with exclusive write access
+        {
+            let mut cached_data = self.cached_data.write().map_err(|_| {
+                GenevaConfigClientError::InternalError("RwLock poisoned".to_string())
+            })?;
+
+            // Double-check in case another thread updated while we were fetching
+            if let Some(existing) = cached_data.token_expiry {
+                if let Some(new) = token_expiry {
+                    if existing >= new {
+                        // Existing token is newer, use it
+                        if let (Some(info), Some(endpoint)) =
+                            (&cached_data.auth_info, &cached_data.token_endpoint)
+                        {
+                            return Ok((info.0.clone(), info.1.clone(), endpoint.clone()));
+                        }
+                    }
+                }
+            }
+
+            // Update cache with fresh data
+            cached_data.auth_info = Some((fresh_token_info, fresh_moniker_info));
+            cached_data.token_endpoint = Some(token_endpoint);
+            cached_data.token_expiry = token_expiry;
+
+            if let (Some(info), Some(endpoint)) =
```

Comment:
Do we need this if else check? At this point, the cache has just been updated under a write lock. Why use the `if let()` check.

Reply:
Good catch, this is not required. Have removed it now.